sn_client/src/audit/dag_crawling.rs

```diff
@@ -62,26 +64,117 @@ impl Client {
         spend_processing: Option<Sender<SignedSpend>>,
         verify: bool,
     ) -> WalletResult<SpendDag> {
-        info!("Building spend DAG from {spend_addr:?}");
-        let mut dag = SpendDag::new(spend_addr);
+        let (tx, mut rx) = tokio::sync::mpsc::channel(SPENDS_PROCESSING_BUFFER_SIZE);
+
+        // start crawling from the given spend address
+        let self_clone = self.clone();
+        let crawl_handle =
+            tokio::spawn(async move { self_clone.spend_dag_crawl_from(spend_addr, tx).await });
+
+        // start DAG building from the spends gathered while crawling
+        // forward spends to processing if provided
+        let build_handle: tokio::task::JoinHandle<Result<SpendDag, WalletError>> =
+            tokio::spawn(async move {
+                let mut dag = SpendDag::new(spend_addr);
+                while let Some(spend) = rx.recv().await {
+                    let addr = spend.address();
+                    dag.insert(addr, spend.clone());
+                    if let Some(sender) = &spend_processing {
+                        sender
+                            .send(spend)
+                            .await
+                            .map_err(|e| WalletError::SpendProcessing(e.to_string()))?;
+                    }
+                }
+                Ok(dag)
+            });
+
+        // wait for both to finish
+        let (crawl_res, build_res) = tokio::join!(crawl_handle, build_handle);
+        crawl_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join crawling results {e}"))
+        })??;
+        let mut dag = build_res.map_err(|e| {
+            WalletError::SpendProcessing(format!("Failed to Join DAG building results {e}"))
+        })??;
+
+        // verify the DAG
+        if verify {
+            info!("Now verifying SpendDAG from {spend_addr:?} and recording errors...");
+            let start = std::time::Instant::now();
+            if let Err(e) = dag.record_faults(&dag.source()) {
+                let s = format!(
+                    "Collected DAG starting at {spend_addr:?} is invalid, this is probably a bug: {e}"
+                );
+                error!("{s}");
+                return Err(WalletError::Dag(s));
+            }
+            let elapsed = start.elapsed();
+            info!("Finished verifying SpendDAG from {spend_addr:?} in {elapsed:?}");
+        }
+
+        Ok(dag)
+    }
+
+    /// Crawls the Spend Dag from a set of given SpendAddresses recursively
```

Comment:
As I understand, for the beta_tracking purpose only, this function shall just:
1, tryto get the spends
2, if failed, put that spendaddress as utxo as well
3, if fetched, put the outputs as utxos
4, return all collected utxos as result
which is somehow a `wide first search` methodology.

Meanwhile, according to the comments and the inner function call `spend_dag_crawl_from` it uses, it seems will tryto drain along one branch first, then go to the next.
which is a `depth first earch` methodology.

I am not sure which one will be better, but my general feeling is that the `wide first` one will be more suitable for the DAG tree we have.

Also, there is `tracking performance` statistics current not being populated in the PR.

I'd strongly suggest to build this auditor separately, then test it against a realy public testnet, to compare the performance with `main`, or even the `PR 1777`, before got it merged.